File: java/2-finish/akka-streams/src/main/java/pipelines/workshop/egress/CustomerSpendLogger.java
package pipelines.workshop.egress;

import akka.NotUsed;
import akka.stream.javadsl.RunnableGraph;
import akka.stream.javadsl.Sink;
import pipelines.akkastream.AkkaStreamlet;
import pipelines.akkastream.javadsl.RunnableGraphStreamletLogic;
import pipelines.streamlets.StreamletShape;
import pipelines.streamlets.avro.AvroInlet;
import pipelines.workshop.schema.scala.CustomerSpendSummary;

public class CustomerSpendLogger extends AkkaStreamlet {

    AvroInlet<CustomerSpendSummary> inlet = AvroInlet.<CustomerSpendSummary>create("in", CustomerSpendSummary.class);

    @Override
    public StreamletShape shape() {
        return StreamletShape.createWithInlets(inlet);
    }

    public RunnableGraphStreamletLogic createLogic() {
        return new RunnableGraphStreamletLogic(getStreamletContext()) {
            @Override
            public RunnableGraph<NotUsed> createRunnableGraph() {
                return getAtMostOnceSource(inlet).to(Sink.foreach(payment -> System.out.println("Customer Spend Summary => " + payment)));
            }
        };
    }
}

File: scala/2-finish/akka-streams/src/main/scala/pipelines/workshop/egress/CustomerSpendLogger.scala
package pipelines.workshop.egress

import akka.stream.scaladsl.RunnableGraph
import pipelines.akkastream.scaladsl.{ FlowWithPipelinesContext, RunnableGraphStreamletLogic }
import pipelines.akkastream.{ AkkaStreamlet, StreamletLogic }
import pipelines.streamlets.StreamletShape
import pipelines.streamlets.avro.AvroInlet
import pipelines.workshop.schema.CustomerSpendAgg

object CustomerSpendLogger extends AkkaStreamlet {

  val inlet = AvroInlet[CustomerSpendAgg]("in")
  val shape = StreamletShape.withInlets(inlet)

  override protected def createLogic(): StreamletLogic = new RunnableGraphStreamletLogic() {

    val flow = FlowWithPipelinesContext[CustomerSpendAgg].map { agg ⇒
      system.log.info(s"$agg")
      agg
    }

    override def runnableGraph(): RunnableGraph[_] =
      atLeastOnceSource(inlet).via(flow).to(atLeastOnceSink)
  }
}
